x/uibc/quota/keeper/quota.go

```diff
@@ -0,0 +1,274 @@
+	var quotaOfIBCDenom uibc.Quota
+
+	if bz == nil {
+		return nil, uibc.ErrNoQuotaForIBCDenom
+	}
+
+	if err := k.cdc.Unmarshal(bz, &quotaOfIBCDenom); err != nil {
+		return nil, err
+	}
+
+	return &quotaOfIBCDenom, nil
+}
+
+// SetQuotaOfIBCDenoms save the rate limits of ibc denoms into store.
+func (k Keeper) SetQuotaOfIBCDenoms(ctx sdk.Context, quotaOfIBCDenoms []uibc.Quota) error {
+	for _, quotaOfIBCDenom := range quotaOfIBCDenoms {
+		if err := k.SetQuotaOfIBCDenom(ctx, quotaOfIBCDenom); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// SetQuotaOfIBCDenom save the quota of ibc denom into store.
+func (k Keeper) SetQuotaOfIBCDenom(ctx sdk.Context, quotaOfIBCDenom uibc.Quota) error {
+	store := ctx.KVStore(k.storeKey)
+	key := uibc.CreateKeyForQuotaOfIBCDenom(quotaOfIBCDenom.IbcDenom)
+
+	bz, err := k.cdc.Marshal(&quotaOfIBCDenom)
+	if err != nil {
+		return err
+	}
+
+	store.Set(key, bz)
+
+	return nil
+}
+
+func (k Keeper) GetTotalOutflowSum(ctx sdk.Context) sdk.Dec {
+	store := ctx.KVStore(k.storeKey)
+	bz := store.Get(uibc.KeyTotalOutflowSum)
+	return sdk.MustNewDecFromStr(string(bz))
+}
+
+// GetQuotaExpires returns ibc-transfer quota expire time.
+func (k Keeper) SetTotalOutflowSum(ctx sdk.Context, amount string) error {
+	store := ctx.KVStore(k.storeKey)
+	key := uibc.KeyTotalOutflowSum
+	store.Set(key, []byte(amount))
+	return nil
+}
+
+// SetQuotaExpires save the quota expires time of ibc denom into store.
+func (k Keeper) SetQuotaExpires(ctx sdk.Context, expires time.Time) error {
+	store := ctx.KVStore(k.storeKey)
+	bz, err := expires.MarshalBinary()
+	if err != nil {
+		return err
+	}
+
+	store.Set(uibc.QuotaExpiresKey, bz)
+
+	return nil
+}
+
+// GetQuotaExpires returns ibc-transfer quota expires time.
+func (k Keeper) GetQuotaExpires(ctx sdk.Context) (*time.Time, error) {
+	store := ctx.KVStore(k.storeKey)
+	bz := store.Get(uibc.QuotaExpiresKey)
+	if bz == nil {
+		return nil, nil
+	}
+	quotaExpires := ctx.BlockTime()
+	if err := quotaExpires.UnmarshalBinary(bz); err != nil {
+		return &quotaExpires, err
+	}
+	return &quotaExpires, nil
+}
+
+// CheckAndUpdateQuota checks the quota of ibc-transfer of denom
+func (k Keeper) CheckAndUpdateQuota(ctx sdk.Context, denom string, sendAmount string) error {
+	params := k.GetParams(ctx)
+
+	quotaOfIBCDenom, err := k.getQuotaOfIBCDenom(ctx, denom)
+	if err != nil {
+		return err
+	}
+
+	if quotaOfIBCDenom == nil {
+		return nil
+	}
+
+	amount, err := strconv.ParseInt(sendAmount, 10, 64)
+	if err != nil {
+		return err
+	}
+
+	// get the registered token settings from leverage
+	tokenSettings, err := k.leverageKeeper.GetTokenSettings(ctx, denom)
+	if err != nil {
+		return nil
+	}
+
+	// get the exchange rate of denom in USD
+	exchangeRate, err := k.oracleKeeper.GetExchangeRate(ctx, denom)
+	if err != nil {
+		return err
+	}
+
+	sendingAmount := sdk.NewDec(amount).Quo(sdk.NewDec(10).Power(uint64(tokenSettings.Exponent)))
+	amountInUSD := exchangeRate.Mul(sendingAmount)
+
```

Comment:
x/oracle doesn't understand ibc/denoms. So if this is "atom", here we will see `ibc/...`, correct? We need to translate it to something like: `atom`.
- We should support `utokens` as well . cc: @toteki have function for that (not sure if `TokenValue()` handles that)

Reply:
`GetExchangeRateBase` will convert `ibc/xxx` to `symbol` and returns the exchange rate in oracle module

Reply:
If supporting uTokens (detect using `leveragetypes.HasUTokenPrefix`) you can use `leveragekeeper.ExchangeUToken` to convert to base tokens then `leveragekeeper.TokenValue` to get USD value.

The last step (`TokenValue`) might behave better for rounding than the math shown above - the reason is `sdk.Dec` having 18-point precision, so `.Quo(sdk.NewDec(10).Power(uint64(tokenSettings.Exponent)` will lose some detail with `exponent >= 18` assets like `DAI`.

Reply:
@gsk967 today the uTokens are enabled over IBC, so we need to handle correct prices.
@toteki do we have an integration test for handling uToken and Token with a set of known prices? Could you point it out to @gsk967 ?

Reply:
If it is `uToken` , We can use above logic of toteki for checking quota of uTokens
https://github.com/umee-network/umee/pull/1568/commits/e485b2ae795163a7f71adcd1d637ddc6c4a924b0#diff-0c1ecbc0c4238830ffea35cdd1e2f3efd4e2d0ec54d4f14e39ea6c3b4abcfac1